277 implementation progress reports

duncandewhurst opened the pull request:
Closes #277 closes #280

duncandewhurst committed:
add forecasts and metrics properties

duncandewhurst committed:
update and run borrow-schema.py, add MetricID codelist

copy Metric definition and Observations. Handle objects without properties. Remove deprecated properties. Fix typos.

duncandewhurst committed:
update changelog

duncandewhurst committed:
Merge branch '0.9-dev' into 277-implementation-progress-reports

duncandewhurst commented:
I'm not sure what is going on with the build failing here, as I can build it OK on my machine.

This PR just addresses the schema, codelist and reference page updates.

Guidance and worked examples will be covered in https://github.com/open-contracting/infrastructure/issues/220 and https://github.com/open-contracting/infrastructure/issues/278 respectively.

jpmckinney commented:
Run `flake8 . --max-line-length 119` locally to reproduce the failures from the Lint workflow.

`metricID.csv` has unquoted commas. Did you write it by hand? It's causing the CI workflow to fail.

duncandewhurst committed:
fix flake8 errors

duncandewhurst committed:
fix MetricID.csv format

duncandewhurst committed:
metricID.csv: capitalize Description

jpmckinney commented on schema/project-level/codelists/metricID.csv:
```diff
@@ -0,0 +1,3 @@
+Code,Title,Description
+financialProgress,Financial Progress,"The planned or actual financial progress for the project. For example, the percentage of spend against the project budget".
```
```suggestion
financialProgress,Financial Progress,"The planned or actual financial progress for the project. For example, the percentage of spend against the project budget."
```

jpmckinney commented on schema/project-level/project-schema.json:
```diff
@@ -1426,6 +1448,141 @@
+              "type": [
+                "string"
+              ],
+              "codelist": "unitClassificationScheme.csv",
```
This codelist doesn't exist in OC4IDS, I don't think, which causes an error.

duncandewhurst replied:
Fixed in https://github.com/open-contracting/infrastructure/pull/282/commits/bed6d33c04fbb933a81863d6521b564863294ec2

duncandewhurst committed:
Update schema/project-level/codelists/metricID.csv

Co-authored-by: James McKinney <26463+jpmckinney@users.noreply.github.com>

duncandewhurst committed:
Add unitClassificationScheme.csv. Fix metricID.csv

duncandewhurst committed:
Merge branch '0.9-dev' into 277-implementation-progress-reports

jpmckinney commented on docs/reference/codelists.md:
```diff
@@ -28,6 +28,7 @@ OC4IDS reuses some codelists from the Open Contracting Data Standard and its ext
 * [Method](https://standard.open-contracting.org/1.1/en/schema/codelists/#method)
 * [Organization identifier scheme](https://standard.open-contracting.org/1.1/en/schema/codelists/#organization-identifier-scheme)
 * [Release tag](https://standard.open-contracting.org/1.1/en/schema/codelists/#release-tag)
+* [Unit Classification Scheme](https://standard.open-contracting.org/1.1/en/schema/codelists/#unit-classification-scheme)
```
Follow the same capitalization as other list items.

```suggestion
* [Unit classification scheme](https://standard.open-contracting.org/1.1/en/schema/codelists/#unit-classification-scheme)
```

jpmckinney commented on docs/reference/changelog.md:
```diff
@@ -4,6 +4,12 @@
 ### Schema
 
+* [#277](https://github.com/open-contracting/infrastructure/issues/277) - add `forecasts` and `metrics` which can be used to publish implementation progress reports
+* [#282](https://github.com/open-contracting/infrastructure/pull/282) - update shared properties to OCDS for PPPs 1.0.0-beta2 / OCDS 1.1.4.
```
```suggestion
* [#282](https://github.com/open-contracting/infrastructure/pull/282) - update fields shared with OCDS for PPPs 1.0.0-beta2 and OCDS 1.1.4.
```

jpmckinney commented on docs/reference/changelog.md:
```diff
@@ -4,6 +4,12 @@
+
+### Codelists
+
+* [#282](https://github.com/open-contracting/infrastructure/pull/282) - update shared codes to OCDS for PPPs 1.0.0-beta2 / OCDS 1.1.4.
```
```suggestion
* [#282](https://github.com/open-contracting/infrastructure/pull/282) - update codes shared with OCDS for PPPs 1.0.0-beta2 and OCDS 1.1.4.
```

jpmckinney commented on schema/borrow-schema.py:
```diff
@@ -333,8 +362,19 @@ def compare(actual, infra_list, ocds_list, prefix, suffix):
+copy_def('Metric', {
+    ('properties', 'id', 'description'): lambda s: s.replace('contracting process', 'contracting process or project')}),  # noqa: E501
+
+schema['definitions']['Metric']['description'] = "Metrics are used to set out forecast and actual metrics targets for a project, for example planned and actual physical and financial progress over time."  # noqa: E501
```
```suggestion
schema['definitions']['Metric']['description'] = "Metrics are used to set out forecast and actual metrics targets for a project: for example, planned and actual physical and financial progress over time."  # noqa: E501
```

jpmckinney commented on schema/project-level/project-schema.json:
```diff
@@ -684,8 +702,7 @@
                   "string"
                 ]
-              },
-              "uniqueItems": true
```
I think OC4IDS had opted-in to more strict validation here.

duncandewhurst replied:
Ah yes, my mistake. Fixed in https://github.com/open-contracting/infrastructure/pull/282/commits/683ef4da0c2d20dfc26e94784e7db0d2fcfeec61

jpmckinney commented on schema/project-level/codelists/metricID.csv:
```diff
@@ -0,0 +1,3 @@
+Code,Title,Description
+financialProgress,Financial Progress,"The planned or actual financial progress for the project. For example, the percentage of spend against the project budget."
+physicalProgress,Physical Progress,"The planned or actual physical progress for a project. For example, the percentage of completion of the physical works."
```
In codelists like documentType we use "Sentence case" instead of "Title Case" for the Title column.

duncandewhurst replied:
Fixed in https://github.com/open-contracting/infrastructure/pull/282/commits/683ef4da0c2d20dfc26e94784e7db0d2fcfeec61

duncandewhurst committed:
Update docs/reference/codelists.md

Co-authored-by: James McKinney <26463+jpmckinney@users.noreply.github.com>

duncandewhurst committed:
Update docs/reference/changelog.md

Co-authored-by: James McKinney <26463+jpmckinney@users.noreply.github.com>

duncandewhurst committed:
Update docs/reference/changelog.md

Co-authored-by: James McKinney <26463+jpmckinney@users.noreply.github.com>

duncandewhurst committed:
Update schema/borrow-schema.py

Co-authored-by: James McKinney <26463+jpmckinney@users.noreply.github.com>

duncandewhurst committed:
restore unique items for gazetteer identifiers, use sentence case in metricID.csv

jpmckinney commented on schema/project-level/project-schema.json:
```diff
@@ -1436,6 +1460,141 @@
+    "Metric": {
+      "type": "object",
+      "title": "Metric",
+      "description": "Metrics are used to set out forecast and actual metrics targets for a project, for example planned and actual physical and financial progress over time.",
```
Need to update to match change in borrow-schema.py (I think running the script again might delete uniqueItems again - maybe easier to just hand-edit this time)

duncandewhurst replied:
Good spot. The output of `borrow-schema.py` had diverged from `project-schema.json` in a couple of other places too, so I've updated the script and run it in https://github.com/open-contracting/infrastructure/pull/282/commits/7fe856f13121ff731d90dd8df57aa0168d7889c0.

duncandewhurst committed:
update and run borrow-schema.py

set stricter validation for gazetteer identifiers and party roles, add people to organizations
